sdks/uniswapx-sdk/src/order/V3DutchOrder.ts

```diff
@@ -0,0 +1,514 @@
+    outputs: V3DutchOutput[];
+};
+
+export type CosignedV3DutchOrderInfo = UnsignedV3DutchOrderInfo & {
+    cosignerData: CosignerData;
+    cosignature: string;
+};
+
+type V3WitnessInfo = {
+    info: OrderInfo,
+    cosigner: string,
+    baseInput: V3DutchInput,
+    baseOutputs: V3DutchOutput[],
+};
+
+
+const COSIGNER_DATA_TUPLE_ABI = "tuple(uint256,address,uint256,uint256,uint256[])";
+
+const V3_DUTCH_ORDER_TYPES = {
+    V3DutchOrder: [
+        { name: "info", type: "OrderInfo" },
+        { name: "cosigner", type: "address" },
+        { name: "baseInput", type: "V3DutchInput" },
+        { name: "baseOutputs", type: "V3DutchOutput[]" },
+    ],
+    OrderInfo: [
+        { name: "reactor", type: "address" },
+        { name: "swapper", type: "address" },
+        { name: "nonce", type: "uint256" },
+        { name: "deadline", type: "uint256" },
+        { name: "additionalValidationContract", type: "address" },
+        { name: "additionalValidationData", type: "bytes" },
+    ],
+    V3DutchInput: [
+        { name: "token", type: "address" },
+        { name: "startAmount", type: "uint256" },
+        { name: "curve", type: "V3Decay" },
+        { name: "maxAmount", type: "uint256" },
+    ],
+    V3DutchOutput: [
+        { name: "token", type: "address" },
+        { name: "startAmount", type: "uint256" },
+        { name: "curve", type: "V3Decay" },
+        { name: "recipient", type: "address" },
+    ],
+    V3Decay: [
+        { name: "relativeBlocks", type: "uint256" },
+        { name: "relativeAmounts", type: "int256[]" },
+    ],
+};
+
+const V3_DUTCH_ORDER_ABI = [
+    "tuple(" + 
+    [
+        "tuple(address,address,uint256,uint256,address,bytes)", // OrderInfo
+        "address", // Cosigner
+        "tuple(address,uint256,tuple(uint256,int256[]),uint256)", // V3DutchInput
+        "tuple(address,uint256,tuple(uint256,int256[]),address)[]", // V3DutchOutput
+        COSIGNER_DATA_TUPLE_ABI,
+        "bytes", // Cosignature
+    ].join(",") + ")",
+];
+
+export class UnsignedV3DutchOrder implements OffChainOrder {
+    public permit2Address: string;
+
+    constructor (
+        public readonly info: UnsignedV3DutchOrderInfo,
+        public readonly chainId: number,
+        _permit2Address?: string
+    ) {
+        this.permit2Address = getPermit2(chainId, _permit2Address);
+    }
+
+    static fromJSON(
+        json: UnsignedV3DutchOrderInfoJSON,
+        chainId: number,
+        _permit2Address?: string
+    ): UnsignedV3DutchOrder {
+        return new UnsignedV3DutchOrder(
+            {
+                ...json,
+                nonce: BigNumber.from(json.nonce),
+                input: {
+                    ...json.input,
+                    startAmount: BigNumber.from(json.input.startAmount),
+                    curve: {
+                        relativeBlocks: json.input.curve.relativeBlocks,
+                        relativeAmounts: json.input.curve.relativeAmounts.map(BigNumber.from),
```

Comment:
hmm i don't think BigNumber supports negative ints?

Reply:
Might have to use BigInt or something

Reply:
Fixed. This was huge, thanks for catching!